Whiskey/Functions/Resolve-WhiskeyDotNetSdkVersion.ps1

```diff
@@ -77,39 +259,236 @@ function Resolve-WhiskeyDotNetSdkVersion
-        if (-not $release)
+        if (-not $release -and $RollForward -eq [Whiskey.DotNetSdkRollForward]::Disable)
         {
             Write-WhiskeyError -Message ('.NET Core release matching "{0}" could not be found in "{1}"' -f $matcher, $releasesIndexUri)
             return
         }
 
-        $releasesJsonUri = $release | Select-Object -ExpandProperty 'releases.json'
-        Write-WhiskeyVerbose -Message ('[{0}] Resolving .NET Core SDK version "{1}" against known released versions at: "{2}"' -f $MyInvocation.MyCommand,$Version,$releasesJsonUri)
+        $searchingMajorMinor = $RollForward -eq [Whiskey.DotNetSdkRollForward]::Major -or $RollForward -eq [Whiskey.DotNetSdkRollForward]::Minor
 
-        $releasesJson = Invoke-RestMethod -Uri $releasesJsonUri -ErrorAction Stop
+        $sdkVersions = $null
+        $resolvedVersion = $null
+        $sdkVersionObjs = $null
+        $searchVersionObj = ConvertTo-VersionObject -VersionToConvert $Version
+        if ( -not (-not $release -and $searchingMajorMinor)) 
+        {
+            $releasesJsonUri = $release | Select-Object -ExpandProperty 'releases.json'
+            Write-WhiskeyVerbose -Message ('[{0}] Resolving .NET Core SDK version "{1}" against known released versions at: "{2}"' -f $MyInvocation.MyCommand,$Version,$releasesJsonUri)
 
-        $sdkVersions = & {
-            $releasesJson.releases |
-                Where-Object { $_ | Get-Member -Name 'sdk' } |
-                Select-Object -ExpandProperty 'sdk' |
-                Select-Object -ExpandProperty 'version'
+            $releasesJson = Invoke-RestMethod -Uri $releasesJsonUri -ErrorAction Stop
 
-            $releasesJson.releases |
-                Where-Object { $_ | Get-Member -Name 'sdks' } |
-                Select-Object -ExpandProperty 'sdks' |
-                Select-Object -ExpandProperty 'version'
+            $sdkVersions = & {
+                $releasesJson.releases |
+                    Where-Object { $_ | Get-Member -Name 'sdk' } |
+                    Select-Object -ExpandProperty 'sdk' |
+                    Select-Object -ExpandProperty 'version'
+
+                $releasesJson.releases |
+                    Where-Object { $_ | Get-Member -Name 'sdks' } |
+                    Select-Object -ExpandProperty 'sdks' |
+                    Select-Object -ExpandProperty 'version'
+            }
+        }
+        if ( $RollForward -and $sdkVersions )
+        {
+            $sdkVersionObjs = $sdkVersions | ForEach-Object { ConvertTo-VersionObject -VersionToConvert $_ }
         }
 
-        $resolvedVersion =
-            $sdkVersions |
-            ForEach-Object { $_ -as [Version] } |
-            Where-Object { $_ -like $Version } |
-            Sort-Object -Descending |
-            Select-Object -First 1
+        switch ($RollForward)
+        {
+            Patch 
+            {
+                $resolvedVersion = 

```

Comment:
We're using a lot of logic and text parsing here with these helper functions to create custom version objects when we could vastly simplify this with using the built in `[System.Version]` type. The .NET SDK version is a 4-part version number (formatted as a 3-part version) and the `[System.Version]` type is also a 4-part version number.

Parse all the available .NET SDK versions as a 4-part version, cast it as a `[Version]` type, and then sort them (.NET knows how to properly sort `[Version]` objects):

```powershell
$sortedSdkVersions =
    $sdkVersions |
    ForEach-Object {
        $_ -match '^(\d+)\.(\d+)\.(\d{1})(\d+)' | Out-Null
        [Version] "$($Matches[1]).$($Matches[2]).$($Matches[3]).$($Matches[4])"
    } |
    Sort-Object -Descending
```

Now in each block of your switch statement you just call `Where-Object` on that list of versions with varying amounts of conditions and then select the first one.


```powershell
switch ($RollForward)
{
    Patch 
    {
        $resolvedVersion = 
            $sortedSdkVersions |
            Where-Object {
                $_.Major -eq $desiredMajor -and
                $_.Minor -eq $desiredMinor -and
                $_.Build -eq $desiredFeature
            } |
            Select-Object -First 1
    }
}
```

Then you can convert that resolved version back into a 3-part version that the dotnet install script expects, remembering to format the patch version with 2 digits.

```powershell
'{0}.{1}.{2}{3:00}' -f $resolvedVersion.Major, $resolvedVersion.Minor, $resolvedVersion.Build, $resolvedVersion.Revision
```

Also, for the rollForward strategies of `patch`, `feature`, `minor`, and `major` [the docs](https://learn.microsoft.com/en-us/dotnet/core/tools/global-json#rollforward) say that these strategies will attempt to use that version if it's installed and will only roll forward as a fallback. So we need to also determine what versions of the .NET SDK are installed globally and use one of those if it matches the given version and if not only then fallback to searching for a newer one.

[Take this code](https://github.com/webmd-health-services/Whiskey/blob/master/Whiskey/Functions/Install-WhiskeyDotNetSdk.ps1#L38-L54) and put it in a new `Get-InstalledDotNetSdk` function that we'll call here to resolve the currently installed versions.